designs/s3/20200210-design-vscode-s3.md

```diff
@@ -0,0 +1,146 @@
+# UX: S3 explorer
+
+* **Type**: UX Design
+* **Status**: Proposal
+
+## Overview
+
+This document proposes a user experience for interacting with AWS S3 from the
+AWS Toolkit ("the Toolkit") for Visual Studio Code ("VSCode").
+
+Some of the features listed below are vague; they are implemented using VSCode
+or platform conventions analogous to the AWS web console.
+
+## Experience
+
+Each region of the Toolkit explorer tree includes an S3 node (see mockup
+below). The toolkit pulls a list of names and properties of S3 objects in the
+currently active profile and presents this data as a tree in the toolkit
+explorer. This is typical of existing toolkit features such as "Schemas" and
+"CDK".
+
+- When a region is expanded, make an eager request to get the count of S3
+  buckets for that region.
+- Show the count of buckets next to the S3 node.
+- The S3 node is always present. If there are no buckets in the region, "grey
+  out" the S3 node.
+- When the current active profile changes, the tree is not refreshed, but is
+  greyed-out to indicate that it is stale. It is not minimized.
+- When a bucket or object is created or deleted, only that node is added or
+  removed from the tree, _without disturbing the current layout of the tree_.
+
+
+![](20200210-design-vscode-s3-tree.png)
+
+### Create and edit S3 buckets
+
+- Root node context menu (right-click):
+  - Create bucket
+- Bucket node context menu (right-click):
+  - Delete bucket
+  - New folder...
+  - Upload...
+  - Copy name
+  - Copy ARN
+
+### Create and edit S3 objects
+
+- S3 object node context menu (right-click):
+  - Delete object
+  - Copy name
+  - Copy ARN
+  - Copy URL
+  - Upload...
+  - Edit
+    - Downloads the object to a temp folder and opens it in a new VSCode editor tab.
+  - Download
+    - Does _not_ present a chooser.
+    - Downloads immediately to the OS default "Downloads" location (typically `~/Downloads`).
+    - Download status is given as a toaster, which contains a "Choose location..." link.
+      - User can change the download location via this link. This will be the
+        default location for future downloads.
+      - The default download location is also configurable as a global Toolkit
+        option. This is not S3-specific, it is used for all "Download"
+        experiences in the Toolkit.
```

Comment:
I don't like this experience, but I'm willing to be convinced otherwise.

* It doesn't make sense to provide users with a way to change their download location after the download has started; could we at the very least move the downloaded file to the selected location?
* I don't think these should be global settings; if you're downloading through the IDE, it makes sense that the download would be to a location that has to do with the current workspace. IIRC Azure downloads to the workspace root.

Reply:
Agree on both points. But that doesn't contradict the rest of the experience. What don't you like?

Reply:
+1 to download-before-changing-location feeling awkward.
Downloading to either the workspace or "downloads" feel appropriate.

Reply:
The main goal is to avoid presenting the user with yet another choice. Download it immediately _somewhere_ reasonable, with the ability to correct it at some time `t` thereafter.

Reply:
If I were the user and the toolkit didn't tell me where this was being downloaded, I would think it's being download to a temp/AppData dir, _a la_ opening something (instead of downloading) through a web browser. 

That would give me a false sense of security that anything I downloaded, no matter how large, would be wiped out automatically at the end of the session, and would have me scrambling to find the file that I just downloaded in case I want to keep it around (something I often do with web browsers...maybe it's just a me problem)

Ultimately, I think we want the downloaded artifact to be prominently visible throughout the whole process; we should show the download progress and the final artifact to remove any ambiguity. Some combination of:

* showing download progress
  * Updating toast (with your call on how much info to show: either a loading bar or an updating percentage, or none at all)
  * Status bar (although I'm not sure how this would surface for multiple objects)
* Showing download location
  * Download straight to workspace dir, so user can see it exists
  * Toast with a link to the file (to open in IDE) and/or a link to open in Explorer/Finder/whatever Linux file browser

I feel that a "download with no explicitly selected location" can work if some or all of that is implemented.

Reply:
> If I were the user and the toolkit didn't tell me where this was being downloaded

The "toaster" will say the download location. Will specify that in the doc.

> either a loading bar or an updating percentage,

Would be nice, perhaps phase-2.

> Toast with a link to the file (to open in IDE)

Definitely 👍

Reply:
> The main goal is to avoid presenting the user with yet another choice. Download it immediately somewhere reasonable, with the ability to correct it at some time t thereafter.

Note that the S3 console actually gives the user two options: *Download* and *Download as..*.

The former is an immediate download to the user's download folder. The latter presents a link that the user can right click to open the browser's download location chooser.

It could be familiar if we had both of those:

`Download` and `Download as...`

That simplifies the implementation and user experience since we no longer need to save a default directory or pop up the toaster with that option.